### PR TITLE
Issue 5

### LINKED ISSUE #5: List subgroups
As a member in a group chat
I want the ability to list all the sub-groups (mini-groups) within the chat
So that I can know the sub-groups within the chat

```
# Command
/list

# Response
[names of subgroups]
```

### PR DESCRIPTION
**Pull request by AshciR**
closes #5 
Screenshot 2024-07-10 at 7 29 41 PM

### CODE REVIEW

File: src/shout_subgroup/create_subgroup.py [hunk 1]
```diff
@@ -1,33 +1,26 @@
 import logging
-import os
```
Comment (AshciR) [inline src/shout_subgroup/create_subgroup.py:2]: IDE removed unneeded imports

File: src/shout_subgroup/create_subgroup.py [hunk 2]
```diff
@@ -1,33 +1,26 @@
...
                           telegram_chat: Chat,
                           subgroup_name: str,
                           usernames: set[str]) -> SubgroupModel:
-    # Telegram uses negative numbers for group chats
-    # If it's a positive number, that means it's an individual.
-    # We can't create subgroup for an individual.
+
     telegram_chat_id = telegram_chat.id
-    if telegram_chat_id >= 0:
+    if not await is_group_chat(telegram_chat_id):
```
Comment (AshciR) [inline src/shout_subgroup/create_subgroup.py:23]: pulled into a util function

File: src/shout_subgroup/list_subgroup.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+from telegram.ext import ContextTypes
+
+from shout_subgroup.database import session
+from shout_subgroup.exceptions import NotGroupChatError
+from shout_subgroup.models import SubgroupModel
+from shout_subgroup.repository import find_all_subgroups_in_group_chat
+from shout_subgroup.utils import is_group_chat
+
+
+async def list_subgroups(db: Session, telegram_group_chat_id: int) -> list[Type[SubgroupModel]]:
```
Comment (AshciR) [inline src/shout_subgroup/list_subgroup.py:15]: Focus of PR

File: src/shout_subgroup/list_subgroup.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        msg = f"Can't list subgroups because telegram chat id {telegram_group_chat_id} is not a group chat."
+        logging.info(msg)
+        raise NotGroupChatError(msg)
+
+    # Get all the subgroups for the group chat
+    subgroups = await find_all_subgroups_in_group_chat(db, telegram_group_chat_id)
+    return subgroups
+
+
+async def list_subgroup_handler(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
```
Comment (AshciR) [inline src/shout_subgroup/list_subgroup.py:27]: Focus of PR

File: src/shout_subgroup/repository.py [hunk 5]
```diff
@@ -1,20 +1,37 @@
...
+async def find_all_users_in_group_chat(db: Session, telegram_group_chat_id: int) -> list[Type[UserModel]]:
     users = db.query(UserModel).join(users_group_chats_join_table).join(GroupChatModel).filter(
         GroupChatModel.telegram_group_chat_id == telegram_group_chat_id
     ).all()
-    
+
     return users
 
+
+async def find_all_subgroups_in_group_chat(db: Session, telegram_group_chat_id: int) -> list[Type[SubgroupModel]]:
```
Comment (AshciR) [inline src/shout_subgroup/repository.py:18]: Focus of PR

File: src/shout_subgroup/shout.py
Comment (AshciR) [inline src/shout_subgroup/shout.py:1]: Pycharm auto formatted this file. We can add a linter to this project during one of our session
https://docs.astral.sh/ruff/

File: tests/test_list_subgroup.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    session.commit()
+
+    yield session
+
+    # Close the session and drop all tables
+    session.close()
+    Base.metadata.drop_all(bind=engine)
+
+
+@pytest.mark.asyncio
```
Comment (AshciR) [inline tests/test_list_subgroup.py:46]: Focus of PR: Added 3 test cases

File: src/shout_subgroup/utils.py [hunk 7]
```diff
@@ -8,3 +7,14 @@ def usernames_valid(usernames: set[str]) -> bool:
     # TODO: Implement stricter validations later
     return all(isinstance(item, str) for item in usernames)
 
+
+async def is_group_chat(telegram_chat_id: int) -> bool:
```
Comment (rick1e) [inline src/shout_subgroup/utils.py:11]: why is this function async?

File: src/shout_subgroup/utils.py [hunk 8]
```diff
@@ -1,5 +1,4 @@
-
-def usernames_valid(usernames: set[str]) -> bool:
+async def usernames_valid(usernames: set[str]) -> bool:
```
Comment (rick1e) [inline src/shout_subgroup/utils.py:1]: why is this async as well?

File: src/shout_subgroup/utils.py (hunk 7 above)
Comment (AshciR) [reply src/shout_subgroup/utils.py:11]: Good question.

I wasn't confident the condition would be checked before the business logic here

```
telegram_chat_id = telegram_chat.id
    if not await is_group_chat(telegram_chat_id):
        msg = f"Can't create subgroup because telegram chat id {telegram_chat_id} is not a group chat."
        logging.info(msg)
        raise NotGroupChatError(msg)

    # If the code reaches to this point the telegram_chat_id has to be a telegram group chat id
    subgroup = await find_subgroup_by_telegram_group_chat_id_and_subgroup_name(db, telegram_chat_id, subgroup_name)
    if subgroup:
        msg = f"Subgroup {subgroup_name} already exists for telegram group chat id {telegram_chat_id}."
        logging.info(msg)
        raise SubGroupExistsError(msg)
```

I erred on the side of caution. If we find out `async` doesn't work this way, let's move it back to sync

File: src/shout_subgroup/utils.py (hunk 8 above)
Comment (AshciR) [reply src/shout_subgroup/utils.py:1]: Similar response as above

Comment (AshciR) [reply src/shout_subgroup/utils.py:1]: If `is_group_chat` is a synchronous function, it could impact the timing and overall performance of the `create_subgroup` function, but the overall effect in terms of logic and functionality would remain the same. Here are the key considerations:

### Synchronous vs. Asynchronous Function Call

1. **Blocking Behavior:**
   - **Synchronous:** The call to `is_group_chat` will block the execution of `create_subgroup` until it completes. This means the entire function will wait for `is_group_chat` to return, potentially causing a delay if `is_group_chat` is slow or resource-intensive.
   - **Asynchronous:** The call to `is_group_chat` would yield control back to the event loop, allowing other tasks to run concurrently while waiting for `is_group_chat` to complete. This can improve the responsiveness and efficiency of your application, especially if `is_group_chat` involves I/O-bound operations like network requests or database queries.

### Overall Effect on `create_subgroup`

1. **Performance:**
   - If `is_group_chat` is a quick operation (e.g., a simple check or a quick database query), the performance difference might be negligible.
   - If `is_group_chat` involves time-consuming operations, making it synchronous can significantly slow down the `create_subgroup` function, especially in a high-load scenario where many such checks are performed concurrently.

2. **Concurrency:**
   - **Asynchronous:** Allows other asynchronous tasks to proceed while waiting for `is_group_chat`, making better use of resources and improving overall throughput.
   - **Synchronous:** Will cause the entire `create_subgroup` function to wait, potentially reducing the efficiency of your application in handling concurrent tasks.

### Practical Example

Consider the following scenario where `is_group_chat` involves a database query:

#### Synchronous Example

```python
def is_group_chat(telegram_chat_id: int) -> bool:
    # Synchronous database query or some other blocking operation
    return db.query(GroupChatModel).filter(GroupChatModel.telegram_group_chat_id == telegram_chat_id).count() > 0

async def create_subgroup(db: Session, telegram_chat: Chat, subgroup_name: str, usernames: set[str]) -> SubgroupModel:
    telegram_chat_id = telegram_chat.id
    if not is_group_chat(telegram_chat_id):  # Synchronous call
        msg = f"Can't create subgroup because telegram chat id {telegram_chat_id} is not a group chat."
        logging.info(msg)
        raise NotGroupChatError(msg)

    # Remaining logic...
```

#### Asynchronous Example

```python
async def is_group_chat(telegram_chat_id: int) -> bool:
    # Asynchronous database query or some other non-blocking operation
    return await db.execute(select(GroupChatModel).filter_by(telegram_group_chat_id=telegram_chat_id)).scalar() is not None

async def create_subgroup(db: Session, telegram_chat: Chat, subgroup_name: str, usernames: set[str]) -> SubgroupModel:
    telegram_chat_id = telegram_chat.id
    if not await is_group_chat(telegram_chat_id):  # Asynchronous call
        msg = f"Can't create subgroup because telegram chat id {telegram_chat_id} is not a group chat."
        logging.info(msg)
        raise NotGroupChatError(msg)

    # Remaining logic...
```

### Conclusion

- **Logic and Functionality:** The overall effect in terms of logic remains the same. The function will correctly check if the chat is a group chat and proceed accordingly.
- **Performance and Concurrency:** Making `is_group_chat` synchronous can impact performance and concurrency. If the operation is quick, the effect may be minimal. However, for slow or I/O-bound operations, asynchronous execution is generally preferred to maintain responsiveness and efficiency.

In summary, the key difference lies in performance and concurrency. The synchronous version may block the execution longer, potentially affecting the application's responsiveness under high load or with slow operations.

**Review by rick1e — APPROVED** [review]
